library/src/main/java/kwasm/format/text/Type.kt

```diff
@@ -14,37 +14,54 @@
 
 package kwasm.format.text
 
+import kwasm.ast.ValueType
 import kwasm.format.ParseContext
+import kwasm.format.ParseException
 
-sealed class Type(
+sealed class Type<T>(
     protected val sequence: CharSequence,
     protected val context: ParseContext? = null
 ) {
+    val value: T by lazy {
+        parseValue()
+    }
 
-    protected abstract fun parseValue(): Unit
+    protected abstract fun parseValue(): T
 
     class ValueType(
         sequence: CharSequence,
         context: ParseContext? = null
-    ) : Type(sequence, context) {
-        override fun parseValue() {
-            TODO("not implemented")
+    ) : Type<kwasm.ast.ValueType>(sequence, context) {
+        override fun parseValue(): kwasm.ast.ValueType {
+            return when (sequence){
+                "i32" -> kwasm.ast.ValueType.I32
+                "i64" -> kwasm.ast.ValueType.I64
+                "f32" -> kwasm.ast.ValueType.F32
+                "f64" -> kwasm.ast.ValueType.F64
+                else -> {
+                    throw ParseException("Invalid ValueType", context)
+                }
+            }
         }
     }
 
     class ResultType(
         sequence: CharSequence,
         context: ParseContext? = null
-    ) : Type(sequence, context) {
-        override fun parseValue() {
-            TODO("not implemented")
+    ) : Type<kwasm.ast.ValueType?>(sequence, context) {
+        override fun parseValue(): kwasm.ast.ValueType? {
+            val keywordAndParameters = getOperationAndParameters(sequence, context)
+            if(keywordAndParameters.first != "result" || keywordAndParameters.second.size > 1){
+                return null
+            }
+            return ValueType(keywordAndParameters.second[0], context).value
```

Comment:
This will throw an exception saying `Invalid ValueType` if it's malformed. That's fine, but I think the context should be shifted to the start of the actual valueType, so the column index in the error message points the user to the right spot.

Reply:
Done